### PR TITLE
Implement YouTube OAuth2 login

### LINKED ISSUE #1588: "Sign in to confirm you’re not a bot." error with YouTube videos
### Bug Description

Playing YouTube videos may fail with the error "Sign in to confirm you’re not a bot."

YouTube seems to have recently rolled out this measure & is affecting all third party players/clients/downloaders. There's currently no known workaround.

### Steps to Reproduce

N/A

### Expected Result

N/A

### Debug Output

N/A

### Additional Info

_No response_

### Checklist

- [X] I have looked for information about this [within the documentation](https://jmusicbot.com/)
- [X] I have searched for similar issues [on the issues page](https://github.com/jagrosh/MusicBot/issues?q=is%3Aissue)
- [X] I am running the latest version of the bot: Release

### LINKED ISSUE #1682: [Bug Report] YT tracks not playing due to 403 error
### Bug Description

so when i go to play music from youtube it gives this error
```
[11:16:44] [ERROR] [AudioHandler]: Track J2X5mJ3HDYE has failed to play
com.sedmelluq.discord.lavaplayer.tools.FriendlyException: Something broke when playing the track.
        at com.sedmelluq.discord.lavaplayer.tools.ExceptionTools.wrapUnfriendlyExceptions(ExceptionTools.java:45)
        at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:121)
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$executeTrack$2(DefaultAudioPlayerManager.java:339)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: Not success status code: 403
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.validateStatusCode(PersistentHttpStream.java:88)
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.attemptConnect(PersistentHttpStream.java:124)
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.connect(PersistentHttpStream.java:106)
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.internalRead(PersistentHttpStream.java:158)
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.read(PersistentHttpStream.java:174)
        at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:295)
        at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275)
        at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaEbmlReader.readEbmlInteger(MatroskaEbmlReader.java:39)
        at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaFileReader.readNextElement(MatroskaFileReader.java:47)
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaStreamingFile.readFile(MatroskaStreamingFile.java:94)
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.loadMatroskaFile(MatroskaAudioTrack.java:52)
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.process(MatroskaAudioTrack.java:35)
        at com.sedmelluq.discord.lavaplayer.track.DelegatedAudioTrack.processDelegate(DelegatedAudioTrack.java:25)
        at dev.lavalink.youtube.track.YoutubeAudioTrack.processStatic(YoutubeAudioTrack.java:148)
        at dev.lavalink.youtube.track.YoutubeAudioTrack.processWithClient(YoutubeAudioTrack.java:116)
        at dev.lavalink.youtube.track.YoutubeAudioTrack.process(YoutubeAudioTrack.java:71)
        at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:109)
        ... 4 common frames omitted
```

### Steps to Reproduce

1.) start the bot
2.) join voice channel
3.) play anything

### Expected Result

music playing

### Debug Output

```
System Properties:
  java.version = 21.0.3
  java.vm.name = OpenJDK 64-Bit Server VM
  java.vm.specification.version = 21
  java.runtime.name = OpenJDK Runtime Environment
  java.runtime.version = 21.0.3+9-LTS
  java.specification.version = 21
  os.arch = aarch64
  os.name = Linux

JMusicBot Information:
  Version = 0.4.3
  Owner = 871764313914871888
  Prefix = @mention
  AltPrefix = !
  MaxSeconds = 0
  NPImages = true
  SongInStatus = false
  StayInChannel = false
  UseEval = false
  UpdateAlerts = true

Dependency Information:
  JDA Version = 4.4.1_353
  JDA-Utilities Version = 3.0.5
  Lavaplayer Version = 2.2.1

Runtime Information:
  Total Memory = 44
  Used Memory = 23

Discord Information:
  ID = 1264486246999265341
  Guilds = 4
  Users = 2
```

### Additional Info

_No response_

### Checklist

- [X] I have looked for information about this [within the documentation](https://jmusicbot.com/)
- [X] I have searched for similar issues [on the issues page](https://github.com/jagrosh/MusicBot/issues?q=is%3Aissue)
- [X] I am using an official release from [the releases page](https://github.com/jagrosh/MusicBot/releases)

### JMusicBot Version

0.4.3

### PR DESCRIPTION
**Pull request by MichailiK**
### This pull request...
  - [ ] Fixes a bug
  - [x] Introduces a new feature
  - [ ] Improves an existing feature
  - [ ] Boosts code quality or performance

### Description
This PR adds a toggle to enable the OAuth2 capability found in the YouTube source manager. It also automatically stores & retrieves the refresh token for convenience of JMusicBot users (as the source manager does not do that itself.)

#### Quick summary of the login flow

Upon enabling OAuth2 (`youtubeoauth2=true` in the config.txt) the YouTube source manager will log the following:

```
[13:41:00] [INFO] [YoutubeOauth2Handler]: ==================================================
[13:41:00] [INFO] [YoutubeOauth2Handler]: !!! DO NOT AUTHORISE WITH YOUR MAIN ACCOUNT, USE A BURNER !!!
[13:41:00] [INFO] [YoutubeOauth2Handler]: OAUTH INTEGRATION: To give youtube-source access to your account, go to https://www.google.com/device and enter code ABC-DEF-GHI
[13:41:00] [INFO] [YoutubeOauth2Handler]: !!! DO NOT AUTHORISE WITH YOUR MAIN ACCOUNT, USE A BURNER !!!
[13:41:00] [INFO] [YoutubeOauth2Handler]: ==================================================
```

Navigating to https://www.google.com/device and entering the code will lead you to an OAuth2 page 

OAuth2 page

(the source manager pretends to be a logged in TV client to play tracks)

Once allowed, the source manager logs a refresh token. For convenience, JMusicBot intercepts the log with a turbo filter & stores the refresh token as a file, `youtubetoken.txt`. The refresh token will be used when JMusicBot gets restarted to retrieve an access token.

### Purpose
This should circumvent the "Sign in to confirm you're not a bot" message that people (particularly) on servers have been getting.

### Relevant Issue(s)
Fixes #1588

Related PR: #1714

### Todo
- [x] The GUI can't seem to log the messages from the OAuth2 handler, they always end up in stdout. Have to yet look into why

### CODE REVIEW

**Comment by chenchuuu** [comment]
It did not works for me. I cloned this branch and after executed run_jmusicbot.sh, in my config.txt did not appear the option youtubeoauth2

**Comment by MichailiK** [comment]
JMusicBot does not alter existing config files. You will need to manually add `youtubeoauth2=true` to your config.txt file.

**Comment by chenchuuu** [comment]
> JMusicBot does not alter existing config files. You will need to manually add `youtubeoauth2=true` to your config.txt file.

I have just did it, but after executed the .sh, youtubesource did not show me this log:

[13:41:00] [INFO] [YoutubeOauth2Handler]: ==================================================
[13:41:00] [INFO] [YoutubeOauth2Handler]: !!! DO NOT AUTHORISE WITH YOUR MAIN ACCOUNT, USE A BURNER !!!
[13:41:00] [INFO] [YoutubeOauth2Handler]: OAUTH INTEGRATION: To give youtube-source access to your account, go to https://www.google.com/device and enter code ABC-DEF-GHI
[13:41:00] [INFO] [YoutubeOauth2Handler]: !!! DO NOT AUTHORISE WITH YOUR MAIN ACCOUNT, USE A BURNER !!!
[13:41:00] [INFO] [YoutubeOauth2Handler]: ==================================================

**Comment by djordjepekic** [comment]
Works great! For anyone wondering, if you run the bot as a service on linux you can use 'sudo journalctl -u <your-service-name>.service -f' to see the logs in real time and to find the code you need to enter into https://www.google.com/device.

**Comment by joseph23pronton** [comment]
i cant build the jar file, how can i build the java file? sorry im new to java things and i usually do php

**Comment by iamayod** [comment]
> i cant build the jar file, how can i build the java file? sorry im new to java things and i usually do php

@joseph23pronton 

I made this a few years ago but try it: https://github.com/iamayod/jmusicbotselfupdateguide

**Comment by Tako88** [comment]
Worked great for me as well!

**Comment by jagrosh** [comment]
I'll need to check the logging issue you described. A few concerns:
- Do these tokens expire? (Do we need to capture the expiration somehow and re-request auth?)
- Does this log message first appear only when trying to play something, or right when the bot starts? (Should we do a dummy request immediately if this feature is enabled?)
- Should we have an enumeration so that owners can opt to receive the request via DM instead of in the logs?

**Comment by MichailiK** [comment]
- They shouldn't expire by themselves (actual TVs using this stay logged in practically forever from my experience), though one likely could manually revoke them through their Google Account.
- Right when the bot starts, more specifically when calling [`yt.useOauth2();`](https://github.com/jagrosh/MusicBot/pull/1670/files#diff-3f5df0ef58313c5ce2c98d41f19e3adb230ec2852fdeb745959addf714a64f53R81)
- I guess we could see if this can be implemented in a neat way

**Comment by jagrosh** [comment]
If there is no expiration - so in theory this only happens once - maybe we just do both (log the message, and attempt to DM the owner).

File: src/main/java/com/jagrosh/jmusicbot/Bot.java [hunk 1]
```diff
@@ -44,6 +45,7 @@ public class Bot
     private final PlaylistLoader playlists;
     private final NowplayingHandler nowplaying;
     private final AloneInVoiceHandler aloneInVoiceHandler;
+    private final YoutubeOauth2TokenHandler youTubeOauth2TokenHandler;
```
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/Bot.java:48]: Honestly a bit of a hacky way of going from the (Lavaplayer) youtube-source logging the details to JDA being ready & DM'ing the owner the details. Feel free to point out better solutions for this!

**Review by MichailiK — COMMENTED** [review]
We probably should (automatically) detect that playback is failing due to the "Sign in to confirm you're not a bot" error and point users to some wiki page on how to sign in

**Comment by IIWDTI** [comment]
I really wish there was a nightly build section, anyone have the latest prerelease version with this fix?

**Comment by shweegl** [comment]
Can someone please build a .jar file for this?

**Comment by MichailiK** [comment]
I've made a PR that adds a new page to the wiki: #1714

Ideally we'll wanna merge this PR & the docs PR roughly at the same time, as I'm referring to the new wiki page in here:

https://github.com/jagrosh/MusicBot/blob/96ffcd8291488ac97b0f25edc958a97986aadcfa/src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java#L210

File: src/main/java/com/jagrosh/jmusicbot/Listener.java [hunk 2]
```diff
@@ -38,14 +40,14 @@
 public class Listener extends ListenerAdapter
 {
     private final Bot bot;
-    
+
```
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/Listener.java]: Undo/get rid of unneccessary diffs present in here & other places in this file

File: src/main/java/com/jagrosh/jmusicbot/Listener.java [hunk 3]
```diff
@@ -85,10 +87,19 @@ public void onReady(ReadyEvent event)
                 catch(Exception ignored) {} // ignored
             }, 0, 24, TimeUnit.HOURS);
         }
+        if (bot.getConfig().useYoutubeOauth2())
+        {
+            YoutubeOauth2TokenHandler.Data data = bot.getYouTubeOauth2Handler().getData();
+            if (data != null)
+            {
+                PrivateChannel channel = bot.getJDA().openPrivateChannelById(bot.getConfig().getOwnerId()).complete();
+                channel.sendMessage("# DO NOT AUTHORISE THIS WITH YOUR MAIN GOOGLE ACCOUNT!!!\n" + "## Create or use an alternative/burner Google account!\n" + "To give JMusicBot access to your Google account, go to " + data.getAuthorisationUrl() + " and enter the code **" + data.getCode() + "**").queue();
```
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/Listener.java]: Formatting could be better

File: src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java [hunk 4]
```diff
@@ -202,8 +204,12 @@ public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason
     }
 
     @Override
-    public void onTrackException(AudioPlayer player, AudioTrack track, FriendlyException exception) {
-        LoggerFactory.getLogger("AudioHandler").error("Track " + track.getIdentifier() + " has failed to play", exception);
+    public void onTrackException(AudioPlayer player, AudioTrack track, FriendlyException exception)
+    {
+        if (exception.getMessage().equals("Sign in to confirm you're not a bot"))
+            LOGGER.error("Track {} has failed to play: {}. You will need to sign in to Google to play YouTube tracks. More info: https://jmusicbot.com/youtube-oauth2", track.getIdentifier(), exception.getMessage());
```
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java]: This could be formatted more nicely

**Comment by OzanKurt** [comment]
Can someone release this build?

**Comment by MichailiK** [comment]
The issue with (enabled) GUI not logging the instructions & showing up in stdout instead is caused by lavalink's youtube source logging the message *before* the GUI gets initialised & intercepts the stdout & stderr

https://github.com/jagrosh/MusicBot/blob/859e5c5862decf433f8face5eaca3372d7d27b22/src/main/java/com/jagrosh/jmusicbot/gui/ConsolePanel.java#L39-L40

A fix for this could be to just initialize the GUI earlier... or we could do nothing as we are also DM'ing the bot owner through Discord.

**Comment by bozoboo** [comment]
Adding youtubeoauth2=true to the config.txt and then starting the bot doesnt do anything. 

The prompt to add the code does not appear.

**Comment by MichailiK** [comment]
It looks like people are still encountering an error stating "Invalid status code for player api response: 400" that I have to yet look into

**Comment by MichailiK** [comment]
I'm uncomfortable merging this at the moment, as YouTube seems to have begun making sign in via OAuth2 ineffective.

**Comment by MichailiK** [comment]
Google probably has rolled out a fix for this workaround to everyone by now.

**Pull request closed without merging**